src/worker/indexers/chainIndexer.ts

```diff
@@ -0,0 +1,285 @@
+import {
+  bulkInsertContractEventLogs,
+  type ContractEventLogEntry,
+} from "../../db/contractEventLogs/createContractEventLogs";
+import { getContractSubscriptionsByChainId } from "../../db/contractSubscriptions/getContractSubscriptions";
+import { getConfig } from "../../utils/cache/getConfig";
+import { getContract } from "../../utils/cache/getContract";
+import { getSdk } from "../../utils/cache/getSdk";
+import { logger } from "../../utils/logger";
+
+export interface GetSubscribedContractsLogsParams {
+  chainId: number;
+  contractAddresses: string[];
+  fromBlock: number;
+  toBlock: number;
+}
+
+/**
+ * Used to abstract eth_getLogs
+ * requirement: throw if any of the rpc calls fail, otherwise some contracts will be out of sync
+ * @param params
+ * @returns ethers.Logs[]
+ */
+export const ethGetLogs = async (params: GetSubscribedContractsLogsParams) => {
+  /* this should use log filter: address: [...contractAddresses] when thirdweb supports */
+  const sdk = await getSdk({ chainId: params.chainId });
+  const provider = sdk.getProvider();
+
+  logger({
+    service: "worker",
+    level: "debug",
+    message: `Starting log fetch - ChainIndexer: ${params.chainId}`,
+  });
+
+  const startTime = performance.now();
+
+  const logs = await Promise.all(
+    params.contractAddresses.map(async (contractAddress) => {
+      const logFilter = {
+        address: contractAddress,
+        fromBlock: ethers.utils.hexlify(params.fromBlock),
+        toBlock: ethers.utils.hexlify(params.toBlock),
+      };
+
+      const logs = await provider.getLogs(logFilter);
+      return logs;
+    }),
+  );
+  const flatLogs = logs.flat();
+
+  const endTime = performance.now();
+  const timeTaken = (endTime - startTime).toFixed(2);
+
+  logger({
+    service: "worker",
+    level: "debug",
+    message: `Completed log fetch - ChainIndexer: ${params.chainId}, received ${flatLogs.length} logs in blocks ${params.fromBlock}-${params.toBlock}, time taken: ${timeTaken} ms`,
+  });
+
+  return flatLogs;
+};
+
+export const getSubscribedContractsLogs = async (
+  params: GetSubscribedContractsLogsParams,
+) => {
+  const sdk = await getSdk({ chainId: params.chainId });
+  const provider = sdk.getProvider();
+
+  // get the log for the contracts
+  const logs = await ethGetLogs(params);
+
+  // cache the contracts and abi
+  const startContractsTime = performance.now();
+  logger({
+    service: "worker",
+    level: "debug",
+    message: `Starting to get contracts - ChainIndexer: ${params.chainId}`,
+  });
+  const uniqueContractAddresses = [
+    ...new Set<string>(logs.map((log) => log.address)),
+  ];
+  const contracts = await Promise.all(
+    uniqueContractAddresses.map(async (address) => {
+      const contract = await getContract({
+        chainId: params.chainId,
+        contractAddress: address,
+      });
+      return { contractAddress: address, contract: contract };
+    }),
+  );
+  const contractCache = contracts.reduce((acc, val) => {
```

Comment:
need to be careful with caching, as a `partial-ABI` can lead to issues. Also, contract ABI/Metadata is cached on by our infra service, namely `contract-api`.

Reply:
this is just locally caching the result of the calls. I guess this line isn't needed, but good to know we cache in sdk. Above should be fine tho